src/storage/src/inner.rs

```diff
@@ -0,0 +1,574 @@
+//
+//     CONSTRAINT TABLES_PRIMARY_KEY
+//         PRIMARY KEY (TABLE_CATALOG, TABLE_SCHEMA, TABLE_NAME),
+//
+//     CONSTRAINT TABLES_FOREIGN_KEY_SCHEMATA
+//         FOREIGN KEY (TABLE_CATALOG, TABLE_SCHEMA)
+//         REFERENCES SCHEMATA,
+//
+//     CONSTRAINT TABLES_CHECK_TABLE_IN_COLUMNS
+//         CHECK (
+//             (TABLE_CATALOG, TABLE_SCHEMA, TABLE_NAME) IN (SELECT TABLE_CATALOG, TABLE_SCHEMA, TABLE_NAME FROM COLUMNS)
+//         ),
+//
+//     CONSTRAINT TABLES_FOREIGN_KEY_USER_DEFINED_TYPES
+//         FOREIGN KEY (USER_DEFINED_TYPE_CATALOG, USER_DEFINED_TYPE_SCHEMA, USER_DEFINED_TYPE_NAME)
+//         REFERENCES USER_DEFINED_TYPES MATCH FULL,
+//
+//     CONSTRAINT TABLES_TYPED_TABLE_CHECK CHECK (
+//         (
+//             IS_TYPED = 'YES'
+//             AND (
+//                 (USER_DEFINED_TYPE_CATALOG,
+//                     USER_DEFINED_TYPE_SCHEMA,
+//                     USER_DEFINED_TYPE_NAME,
+//                     SELF_REFERENCING_COLUMN_NAME,
+//                     REFERENCE_GENERATION) IS NOT NULL
+//                 AND
+//                 (TABLE_CATALOG, TABLE_SCHEMA, TABLE_NAME, SELF_REFERENCING_COLUMN_NAME) IN
+//                     (SELECT TABLE_CATALOG, TABLE_SCHEMA, TABLE_NAME, COLUMN_NAME FROM COLUMNS WHERE IS_SELF_REFERENCING = 'YES')
+//             )
+//         ) OR (
+//             IS_TYPED = 'NO'
+//             AND (
+//                 USER_DEFINED_TYPE_CATALOG,
+//                 USER_DEFINED_TYPE_SCHEMA,
+//                 USER_DEFINED_TYPE_NAME,
+//                 SELF_REFERENCING_COLUMN_NAME,
+//                 REFERENCE_GENERATION ) IS NULL
+//         )
+//     ),
+//
+//     CONSTRAINT TABLES_SELF_REFERENCING_COLUMN_CHECK
+//         CHECK (
+//             (SELF_REFERENCING_COLUMN_NAME, REFERENCE_GENERATION) IS NULL
+//             OR (SELF_REFERENCING_COLUMN_NAME, REFERENCE_GENERATION) IS NOT NULL
+//         ),
+//
+//     CONSTRAINT TABLES_TEMPORARY_TABLE_CHECK
+//         CHECK (
+//             (TABLE_TYPE IN ('GLOBAL TEMPORARY', 'LOCAL TEMPORARY')) = (COMMIT_ACTION IS NOT NULL)
+//         ),
+//
+//     CONSTRAINT TABLES_CHECK_NOT_VIEW
+//         CHECK (
+//             NOT EXISTS (
+//                 SELECT  TABLE_CATALOG, TABLE_SCHEMA, TABLE_NAME
+//                 FROM    TABLES
+//                 WHERE   TABLE_TYPE = 'VIEW'
+//                 EXCEPT
+//                 SELECT  TABLE_CATALOG, TABLE_SCHEMA, TABLE_NAME
+//                 FROM    VIEWS
+//             )
+//         )
+// )
+const TABLES_TABLE: &'_ str = "TABLES";
+
+#[derive(Debug, PartialEq)]
+pub enum StorageError {
+    RuntimeCheckError,
+    SystemError(SystemError),
+}
+
+pub type StorageResult<T> = std::result::Result<T, StorageError>;
+
+type RecordId = u8;
+
+pub struct Storage {
+    sequences: RwLock<HashMap<String, u8>>,
+    schemas: RwLock<HashMap<String, u8>>,
+    tables: RwLock<HashMap<String, u8>>,
+    on_disk: Option<OnDiskStorage>,
```

Comment:
Is there a use case where we wouldn't use OnDiskStorage?

How are the u8's meant to be used by the rest of the database?

Reply:
> Is there a use case where we wouldn't use OnDiskStorage?

Yes, unit tests for `sql_engine`

> How are the u8's meant to be used by the rest of the database?

I am reworking the whole code from the PR. There wouldn't be `Frontend` and `Backend` storage. I come to realisation that `Frontend` is actually `DatabaseManager` inside `sql_engine`, so I moving all of it there. `Backend` is going to be a thin wrapper around `sled` just to abstract away dependency. I am going to change `u8` to `u64`. For system schemas and tables they will form their Ids. For non system tables they are going to be `record_id`s - keys for `sled` trees.

> I come to realisation that `Frontend` is actually `DatabaseManager` inside `sql_engine`

That should help with #114 and with implementing indexes. `sled` keys will be indexes columns and `record_id` - value, so by looking up into index you know what record you want to read